### PR TITLE
Add TraceEnumSample_ELBO = TraceEnum_ELBO + infer_discrete

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #1824 

This adds a `TraceEnumSample_ELBO` class that allows sharing work between `TraceEnum_ELBO` and a subsequent call to `infer_discrete`, which can speed up inference by over ~1.6x. From the docs:

> The following are equivalent but the first is cheaper, sharing work
between the computations of ``loss`` and ``z``:
```py
# Version 1.
elbo = TraceEnumSample_ELBO(max_plate_nesting=1)
loss = elbo.loss(*args, **kwargs)
z = elbo.sample_saved()

# Version 2.
elbo = TraceEnum_ELBO(max_plate_nesting=1)
loss = elbo.loss(*args, **kwargs)
guide_trace = poutine.trace(guide).get_trace(*args, **kwargs)
z = infer_discrete(poutine.replay(model, guide_trace),
                   first_available_dim=-2)(*args, **kwargs)
```

This PR also makes `infer_discrete` compatible with data subsampling, switching from `log_prob` to `unscaled_log_prob` to match behavior in `TraceEnum_ELBO`.

## Tested
- added unit tests
- tested extensively on the contrib-treecat branch with large CPU and GPU computations

### CODE REVIEW

**Comment by fritzo** [comment]
ready to merge?
